recipe/patches/0010-don-t-declare-RowMajorStrg-as-extern-on-windows.patch

```diff
@@ -0,0 +1,122 @@
+From 729749ad7fa884a233cffb9c242e70975d5c65aa Mon Sep 17 00:00:00 2001
+From: "H. Vetinari" <h.vetinari@gmx.com>
+Date: Sat, 26 Oct 2024 20:13:34 +1100
+Subject: [PATCH 10/12] don't declare RowMajorStrg as extern on windows
+
+on platforms with HAS_ATTRIBUTE_WEAK_SUPPORT, this presumably plays a role in
+the linker shenanigans with cblas_xerbla, but on windows it just blows up
```

Comment:
Without any patch, it runs into 
```
[9764/9794] Linking Fortran executable bin\xscblat2.exe
FAILED: bin/xscblat2.exe 
C:\Windows\system32\cmd.exe /C "cd . && %BUILD_PREFIX%\Library\bin\cmake.exe -E vs_link_exe --intdir=CBLAS\testing\CMakeFiles\xscblat2.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- %BUILD_PREFIX%\Library\bin\lld-link.exe /nologo CBLAS\testing\CMakeFiles\xscblat2.dir\c_sblat2.f.obj CBLAS\testing\CMakeFiles\xscblat2.dir\c_sblas2.c.obj CBLAS\testing\CMakeFiles\xscblat2.dir\c_s2chke.c.obj CBLAS\testing\CMakeFiles\xscblat2.dir\auxiliary.c.obj CBLAS\testing\CMakeFiles\xscblat2.dir\c_xerbla.c.obj  /out:bin\xscblat2.exe /implib:lib\xscblat2.lib /pdb:bin\xscblat2.pdb /version:0.0 /machine:x64  -Wl,-defaultlib:D:/bld/blas-split_1729934591212/_build_env/lib/clang/19/lib/windows/clang_rt.builtins-x86_64.lib /INCREMENTAL:NO /subsystem:console  lib\cblas.lib -libpath:"D:/bld/blas-split_1729934591212/_build_env/Library/lib" -libpath:"D:/bld/blas-split_1729934591212/_build_env/Library/lib/clang/19/lib/windows"  && cd ."
LINK: command "%BUILD_PREFIX%\Library\bin\lld-link.exe /nologo CBLAS\testing\CMakeFiles\xscblat2.dir\c_sblat2.f.obj CBLAS\testing\CMakeFiles\xscblat2.dir\c_sblas2.c.obj CBLAS\testing\CMakeFiles\xscblat2.dir\c_s2chke.c.obj CBLAS\testing\CMakeFiles\xscblat2.dir\auxiliary.c.obj CBLAS\testing\CMakeFiles\xscblat2.dir\c_xerbla.c.obj /out:bin\xscblat2.exe /implib:lib\xscblat2.lib /pdb:bin\xscblat2.pdb /version:0.0 /machine:x64 -Wl,-defaultlib:D:/bld/blas-split_1729934591212/_build_env/lib/clang/19/lib/windows/clang_rt.builtins-x86_64.lib /INCREMENTAL:NO /subsystem:console lib\cblas.lib -libpath:D:/bld/blas-split_1729934591212/_build_env/Library/lib -libpath:D:/bld/blas-split_1729934591212/_build_env/Library/lib/clang/19/lib/windows /MANIFEST:EMBED,ID=1" failed (exit code 1) with the following output:
lld-link: warning: ignoring unknown argument '-Wl,-defaultlib:D:/bld/blas-split_1729934591212/_build_env/lib/clang/19/lib/windows/clang_rt.builtins-x86_64.lib'

lld-link: error: undefined symbol: RowMajorStrg

>>> referenced by CBLAS\testing\CMakeFiles\xscblat2.dir\c_xerbla.c.obj:($LN60)
```

When I tried to add `__declspec(selectany)` there, I [ran into](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1065120&view=logs&j=fe7faf59-69bb-5487-b8fd-a35a7bda3a51&t=6610678e-79db-5ced-d3c4-e5157de7c8b6)
```
[8444/9794] Building C object CBLAS\testing\CMakeFiles\xscblat2.dir\c_s2chke.c.obj
FAILED: CBLAS/testing/CMakeFiles/xscblat2.dir/c_s2chke.c.obj 
C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\cl.exe  /nologo  -I%SRC_DIR%\CBLAS\include -I%SRC_DIR%\build\include /DWIN32 /D_WINDOWS /W3 /MD /O2 /Ob2 /DNDEBUG /showIncludes /FoCBLAS\testing\CMakeFiles\xscblat2.dir\c_s2chke.c.obj /FdCBLAS\testing\CMakeFiles\xscblat2.dir\ /FS -c %SRC_DIR%\CBLAS\testing\c_s2chke.c
%SRC_DIR%\CBLAS\testing\c_s2chke.c(34): error C2205: 'RowMajorStrg': cannot initialize extern variables with block scope
```

The thing is that there isn't really a point (AFAICT) for having `RowMajorStrg` be `extern`, because it gets [overridden](https://github.com/Reference-LAPACK/lapack/blob/v3.9.0/CBLAS/testing/c_c2chke.c#L54) for each test _in the same translation unit_. That's why I went with a direct initialization (except in xerbla itself, where we do need it to be `extern`).

Reply:
> The thing is that there isn't really a point (AFAICT) for having RowMajorStrg be extern, because it gets [overridden](https://github.com/Reference-LAPACK/lapack/blob/v3.9.0/CBLAS/testing/c_c2chke.c#L54) for each test in the same translation unit.

There is a point. It sets the value of `RowMajorStrg` to be used by `cblas_xerbla`. With your implementation `cblas_xerbla` does not see that value.

Reply:
The unpatched sources failed to compile. I'm happy to do this differently, but out of several attempts, these two patches were the only two that I got working.

My naïve understand was that `__declspec(selectany)` in `cblas_xerbla` will get the current value of `RowMajorStrg`, but admittedly, I had never seen much less used that attribute before researching the failure here.

OTOH, the test suite passes, and I'd be kinda surprised if that was possible without xerbla getting the right `RowMajorStrg`... 🤔

Reply:
> My naïve understand was that __declspec(selectany) in cblas_xerbla will get the current value of RowMajorStrg

If all the variables are local in other functions, how can `cblas_xerbla` use those?

> OTOH, the test suite passes, and I'd be kinda surprised if that was possible without xerbla getting the right RowMajorStrg... 

The tests are mostly skipped on windows, so it doesn't tell us anything.

Reply:
`xscblat2` (which causes the first build failure if unpatched) is definitely being run:
```
        Start  13: CBLAS-xscblat1
 12/121 Test  #12: BLAS-xblat3z .....................   Passed    0.03 sec
        Start  14: CBLAS-xscblat2
 13/121 Test  #13: CBLAS-xscblat1 ...................   Passed    0.01 sec
        Start  15: CBLAS-xscblat3
 14/121 Test  #14: CBLAS-xscblat2 ...................   Passed    0.04 sec
        Start  16: CBLAS-xdcblat1
 15/121 Test  #16: CBLAS-xdcblat1 ...................   Passed    0.01 sec
        Start  17: CBLAS-xdcblat2
 16/121 Test  #15: CBLAS-xscblat3 ...................   Passed    0.05 sec
        Start  18: CBLAS-xdcblat3
 17/121 Test  #17: CBLAS-xdcblat2 ...................   Passed    0.04 sec
        Start  19: CBLAS-xccblat1
 18/121 Test  #19: CBLAS-xccblat1 ...................   Passed    0.01 sec
        Start  20: CBLAS-xccblat2
 19/121 Test  #18: CBLAS-xdcblat3 ...................   Passed    0.07 sec
        Start  21: CBLAS-xccblat3
 20/121 Test  #20: CBLAS-xccblat2 ...................   Passed    0.05 sec
        Start  22: CBLAS-xzcblat1
 21/121 Test  #22: CBLAS-xzcblat1 ...................   Passed    0.01 sec
        Start  23: CBLAS-xzcblat2
 22/121 Test  #21: CBLAS-xccblat3 ...................   Passed    0.07 sec
        Start  24: CBLAS-xzcblat3
 23/121 Test  #23: CBLAS-xzcblat2 ...................   Passed    0.05 sec
```

I've traced a couple others of the implementation files touched by the patch
```
 CBLAS/testing/c_c2chke.c | 2 +-
 CBLAS/testing/c_c3chke.c | 2 +-
 CBLAS/testing/c_d2chke.c | 2 +-
 CBLAS/testing/c_d3chke.c | 2 +-
 CBLAS/testing/c_s2chke.c | 2 +-
 CBLAS/testing/c_s3chke.c | 2 +-
 CBLAS/testing/c_z2chke.c | 2 +-
 CBLAS/testing/c_z3chke.c | 2 +-
```
to the test binaries they appear in, and they all get run AFAICT.

Reply:
> If all the variables are local in other functions, how can `cblas_xerbla` use those?

That'd be down to whatever magic `__declspec(selectany)` does. 😅 

Without that, we predictably fail with linker errors. Not sure what happens exactly if specified, but the place where I [found](https://stackoverflow.com/a/2290838) this mentions it specifically covers part of the functionality of weak symbols (which lapack is using if available, hence on unix).

Reply:
I mean, if the choice is between ignoring/skipping the tests to avoid the patch, or patching purely test-related files (under `CBLAS/testing`) and running+passing the tests, I prefer the latter -- even if the patch is perhaps not perfect.

Again, happy to improve this if you have suggestions how to fix the compilation errors. Otherwise I think this should be OK to move forward.

Reply:
Use the following patch
```diff
diff --git a/CBLAS/include/cblas.h b/CBLAS/include/cblas.h
index b8baf4eca..a830d0644 100644
--- a/CBLAS/include/cblas.h
+++ b/CBLAS/include/cblas.h
@@ -4,6 +4,16 @@
 #include <stdint.h>
 #include <inttypes.h>
 
+#ifdef _MSC_VER
+#ifdef CBLAS_BUILDING_DLL
+#define CBLAS_DLL __declspec(dllexport)
+#else
+#define CBLAS_DLL __declspec(dllimport)
+#endif
+#else
+#define CBLAS_DLL
+#endif
+
 
 #ifdef __cplusplus
 extern "C" {            /* Assume C declarations for C++ */
diff --git a/CBLAS/src/CMakeLists.txt b/CBLAS/src/CMakeLists.txt
index 8dcb2f293..833311e3f 100644
--- a/CBLAS/src/CMakeLists.txt
+++ b/CBLAS/src/CMakeLists.txt
@@ -122,6 +122,9 @@ set_target_properties(${CBLASLIB}_obj PROPERTIES POSITION_INDEPENDENT_CODE ON)
 if(HAS_ATTRIBUTE_WEAK_SUPPORT)
   target_compile_definitions(${CBLASLIB}_obj PRIVATE HAS_ATTRIBUTE_WEAK_SUPPORT)
 endif()
+if(BUILD_SHARED_LIBS)
+  target_compile_definitions(${CBLASLIB}_obj PRIVATE CBLAS_BUILDING_DLL)
+endif()
 
 if(BUILD_INDEX64_EXT_API)
   # 64bit Integer Interface
diff --git a/CBLAS/src/cblas_globals.c b/CBLAS/src/cblas_globals.c
index 5d91a18c9..b1febc408 100644
--- a/CBLAS/src/cblas_globals.c
+++ b/CBLAS/src/cblas_globals.c
@@ -1,2 +1,8 @@
-   int CBLAS_CallFromC=0;
-   int RowMajorStrg=0;
+#ifdef _MSC_VER
+#define CBLAS_DLL __declspec(dllexport)
+#else
+#define CBLAS_DLL
+#endif
+
+CBLAS_DLL int CBLAS_CallFromC=0;
+CBLAS_DLL int RowMajorStrg=0;
```

and do a `sed -i.bak "s/extern int RowMajorStrg/CBLAS_DLL extern int RowMajorStrg/g` and similar for `CBLAS_CallFromC` in all the CBLAS sources.

Reply:
Thank you!